### PR TITLE
Recording of robotics nodes

### PR DESCRIPTION
**Pull request by h3ndrk**
## Introduced Changes

- `PersistentState` is renamed to `CyclerState` because it stores state in a cycler
- Node states, setup main outputs, and cross-inputs (from other cyclers) are recorded by serializing to Bincode and writing it to disk
- `hardware.json` is split into `framework.json` and `hardware.json`
- Recording can be enabled by changing the `framework.json`

## ToDo / Known Issues

- Naming of the structs?
- Fix Pepsi parameter setting

## Ideas for Next Iterations (Not This PR)

- Delete localization recording
- Change pixel storage
- Reuce frequency
- Same seconds in `launchHULK` log files and recording file names
- Reduce recording interleavement
- Add channel directly from robotics domain to "runtime" framework

## How to Test

Deploy to a NAO, check that the cyclers are enabled for recording. When the NAO is in playing primary state, the NAO records to `logs` directory in Bincode format.

### CODE REVIEW

File: crates/context_attribute/src/lib.rs [hunk 1]
```diff
@@ -91,13 +94,13 @@ pub fn context(_attributes: TokenStream, input: TokenStream) -> TokenStream {
...
                     },
-                    "Parameter" | "PersistentState" => match &mut first_segment.arguments {
+                    "CyclerState" | "Parameter" => match &mut first_segment.arguments {
                         PathArguments::AngleBracketed(arguments) if arguments.args.len() == 2 => {
                             pop_string_argument(arguments);
                             let data_type = get_data_type(arguments);
                             into_reference_with_lifetime(
                                 data_type,
-                                (first_segment.ident == "PersistentState").then(Mut::default),
+                                (first_segment.ident != "Parameter").then(Mut::default),
```
Comment (knoellle) [inline crates/context_attribute/src/lib.rs:103]: Why did you change this comparison?
Previously it said that PersistentState requires a mutable reference. Now it says that parameters don't.
But from just reading the code, no statement is made about whether the Persistent-/CyclerState require it, that information is implicit.

**Review by knoellle — CHANGES_REQUESTED** [review]

**Review by knoellle — COMMENTED** [review]
Looks good so far. I have left some comments.
From what I can see, the main things left to implement are:
 - serializing node state
 - writing frames to a file

**Comment by schmidma** [comment]
Did I miss something? I can't find the `NodeState` anywhere?

**Comment by knoellle** [comment]
There is no `NodeState` identifier.
All member field of the node structs are part of the node state.
However, fields that are skipped during (de)serialization must be considered ephemeral and should only be used for caching.

**Comment by schmidma** [comment]
As you are implementing `Serialize` and `Deserialize` for nearly all nodes, is it possible to expose these to communication now, such that they can be subscribed via twix?

**Comment by knoellle** [comment]
Not without further changes, no.
I played around with it for a bit and I don't see a way to construct an AdditionalOutput with a type path that is valid both in the module itself as well as in the generated structs module for the database types.
The only way I see for this to work is to move the state to a separate crate (as we do with `types`), but that's no different from the state on the main branch.

**Comment by schmidma** [comment]
Sorry, my previous statement was phrased in a confusing manner. I was expecting further necessary changes. The question was meant in general. We introduce a new (very strong) requirement to nodes: implementing `Serialize` and `Deserialize`. We previously decided against such requirements. But introducing these implementations, a node can directly serialized and thus also exposed via communication.

To realize such a thing, I imagine the generated framework to expose not only databases containing the main and additional outputs, but also the list of nodes to communication. I think they do not necessarily need to exist in a separate crate for that, do they?

Maybe we should discuss further details after the upcoming dev meeting or at another occasion.

**Comment by knoellle** [comment]
Yes, introducing this trait bound certainly opens up the possibility of exposing all nodes state via communication.
Although we also have to consider `SerializeHierarchy`.

File: crates/code_generation/src/cyclers.rs [hunk 2]
```diff
@@ -438,6 +447,139 @@ fn generate_cycle_method(cycler: &Cycler, cyclers: &Cyclers) -> TokenStream {
...
+                        ..
+                    }
+            ) && !required_inputs.contains(field)
+            {
+                required_inputs.push(field.clone());
+            }
+        }
+    }
+    required_inputs
+}
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs:473]: It's a naming thing. I named it like that because these are all needed inputs for executing the cycle in replay.

File: crates/code_generation/src/cyclers.rs [hunk 3]
```diff
@@ -415,6 +420,9 @@ fn generate_cycle_method(cycler: &Cycler, cyclers: &Cyclers) -> TokenStream {
                     own_database.deref_mut()
                 };
 
+                let enable_recording = <HardwareInterface as hardware::RecordingInterface>::get_recording(&*self.hardware_interface);
+                let mut recording_frame = Vec::new(); // TODO: possible optimization: cache capacity
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs:424]: It seems, I forgot to serialize the nodes. The idea was to do this at the beginning. This would be the information needed for replay. Currently, I would try a simple implementation by serializing every frame. Let's see, if we need to optimize.

File: tools/behavior_simulator/src/interfake.rs [hunk 4]
```diff
@@ -23,6 +23,14 @@ impl NetworkInterface for Interfake {
     }
 }
 
+impl RecordingInterface for Interfake {
+    fn get_recording(&self) -> bool {
+        false
+    }
+
+    fn set_recording(&self, _enable: bool) {}
+}
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply tools/behavior_simulator/src/interfake.rs:32]: We need to discuss these interesting ideas. For the current maturity level, my goal is a feasibility/performance test. Depending on the result we can think about the next steps.

**Comment by h3ndrk** [comment]
> Did I miss something? I can't find the `NodeState` anywhere?

Sorry, please ignore outdated the description... 😅

**Comment by h3ndrk** [comment]
> As you are implementing `Serialize` and `Deserialize` for nearly all nodes, is it possible to expose these to communication now, such that they can be subscribed via twix?

In theory yes but out-of-scope for this PR.

**Comment by h3ndrk** [comment]
> Not without further changes, no.
> 
> I played around with it for a bit and I don't see a way to construct an AdditionalOutput with a type path that is valid both in the module itself as well as in the generated structs module for the database types.
> 
> The only way I see for this to work is to move the state to a separate crate (as we do with `types`), but that's no different from the state on the main branch.

If I would implement it, I would build a separate Communication infrastructure and don't rely on e.g. additional outputs. But this needs more design work.

**Comment by h3ndrk** [comment]
> Yes, introducing this trait bound certainly opens up the possibility of exposing all nodes state via communication.
> 
> Although we also have to consider `SerializeHierarchy`.

Exactly 100% what Schmiddy wrote above. But I think we don't need a trait, Code generation should be enough. But it needs more thinking.

File: crates/code_generation/src/cyclers.rs [hunk 5]
```diff
@@ -438,6 +447,139 @@ fn generate_cycle_method(cycler: &Cycler, cyclers: &Cyclers) -> TokenStream {
...
+        };
+        quote! {
+            if enable_recording {
+                bincode::serialize_into(&mut recording_frame, #value_to_be_recorded).wrap_err(#error_message)?;
+            }
+        }
+    });
+    quote! {
+        #(#recordings)*
+    }
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs]: Done

**Comment by h3ndrk** [comment]
The current approach has been tested on a real NAO. Here are the findings and conclusions:

# Size of recorded data

1 - 3 GiB data is written per minute. 10 minutes of one half time would be 10 - 30 GiB. The NAO has 32 GiB internal memory. USB sticks can be used for external storage, our usual sticks have 32 GiB storage. That means we might run into storage size problems during games.

One possibility is to optimize the amount of data further but it is unlikely that this will reduce data usage by orders of magnitude. Only some cyclers accumulate a lot of data (vision) and the serialization only has 30 % overhead compared to the raw data. Compression is unlikely to help because this is mostly image data which has high entropy and is hard to compress losslessy. Another possibility is to reduce the frequency of recorded frames, maybe even for some cyclers while recording other cyclers at full frequency.

# Execution time impact

The biggest impact has vision recording with 20 ms Bincode serialization and writing to `BufWriter<File>`. This leaves 13 ms headroom for the actual calculation which is nearly the time our current vision pipeline needs for calculation. Control cycler recording takes 0.3 ms and the whole cycle takes 2 ms (10 ms headroom). The other cyclers are even faster. Again, vision is the critical path and needs special care. If we would reduce the frequency, only every other frame would take longer. Maybe we would skip images but we control that via the frequency. One idea is to change the internal buffer format of `YCbCr422Image` to `Vec<u8>` to decrease one layer of serialization hopefully resulting in simple generated code. All accesses must be changed to use the different buffer format. Maybe we need to introduce some `unsafe` stuff to efficiently cast pointers around.

**Comment by h3ndrk** [comment]
Some ideas to support different frequencies and dynamic recording.

Dynamic recording may be triggered by different parties. For example a trigger might be the current game state (robotics domain) or someone requesting a recording from Twix (debug tooling, framework domain via communication) or it has been requested from the deployed configuration (framework domain via communication).

Communication and the hardware interface will both provide a quadruple state: Force to not record, force to record, intend to not record, intend to record. This is also the order of descending precendence. Communication will provide a state for each cycler enabling fine-grained control of the recording. The hardware interface may just provide a single state for the moment, but this can be extended later to one for each cycler.

If the cycler derives from all states that a recording of the current frame should happen, it records everything and sends the data to its instance of a channel to communication. Communication will receive the data and store it in the file system. We might think about, reducing computation overhead in the cyclers even more by moving the serialization into communication.

Communication will also provide recording frequency information, for example whether every frame should be recorded, none at all or at which frequency. One approach is to have this fixed over the whole process runtime by storing this information in e.g. the `hardware.json` configuration and loading it once at startup. Another approach is to also provide this information via communication and be able to control it from a connected debug client at runtime. Dynamic frequencies have the benefit of more fine-grained analysis during testing without restarting the HULK process.

Communication will serve an additional API endpoint for recording s.t. clients are able to control the recording state and maybe also the frequency.

File: crates/code_generation/src/cyclers.rs [hunk 6]
```diff
@@ -353,11 +371,13 @@ fn generate_cycle_method(cycler: &Cycler, cyclers: &Cyclers) -> TokenStream {
...
         .setup_nodes
         .iter()
-        .map(|node| generate_node_execution(node, cycler));
+        .map(|node| generate_node_execution(node, cycler, true));
     let cycle_node_executions = cycler
         .cycle_nodes
         .iter()
-        .map(|node| generate_node_execution(node, cycler));
+        .map(|node| generate_node_execution(node, cycler, false));
+    let required_inputs = get_required_inputs(cycler);
```
Comment (h3ndrk) [inline crates/code_generation/src/cyclers.rs:379]: Unused?

File: crates/code_generation/src/cyclers.rs [hunk 7]
```diff
@@ -415,29 +446,178 @@ fn generate_cycle_method(cycler: &Cycler, cyclers: &Cyclers) -> TokenStream {
...
                     #lock_readers
+                    #required_input_recordings
                     #(#cycle_node_executions)*
                 }
 
                 #after_remaining_nodes
+
+                let cycle_end = cycle_begin.elapsed();
+                if enable_recording {
+                    println!("{instance:?}\t{cycle_end:?}\t{recording_duration:?}");
```
Comment (h3ndrk) [inline crates/code_generation/src/cyclers.rs:474]: Remove debug prints

File: crates/code_generation/src/cyclers.rs [hunk 8]
```diff
@@ -451,17 +631,23 @@ fn generate_perception_cycler_updates(cyclers: &Cyclers) -> TokenStream {
...
+    let recording_error_message = format!("failed to record `{}`", node.name);
+    let cycle_error_message = format!("failed to execute cycle of `{}`", node.name);
+    let database_updates = generate_database_updates(node, generate_recording);
     let database_updates_from_defaults = generate_database_updates_from_defaults(node);
     quote! {
         {
+            if enable_recording {
+                let begin = std::time::Instant::now();
+                bincode::serialize_into(&mut recording_frame, &self.#node_member).wrap_err(#recording_error_message)?;
+                recording_duration += begin.elapsed();
```
Comment (h3ndrk) [inline crates/code_generation/src/cyclers.rs:649]: Remove execution time measurements

File: crates/code_generation/src/cyclers.rs [hunk 9]
```diff
@@ -698,20 +884,28 @@ fn generate_context_initializers(node: &Node, cycler: &Cycler) -> TokenStream {
...
-            })
-    ;
+            });
     quote! {
         #(#initializers,)*
     }
 }
 
-fn generate_database_updates(node: &Node) -> TokenStream {
+fn generate_database_updates(node: &Node, generate_recording: bool) -> TokenStream {
```
Comment (h3ndrk) [inline crates/code_generation/src/cyclers.rs:893]: Enum

File: crates/control/src/localization_recorder.rs [hunk 10]
```diff
@@ -15,7 +15,9 @@ use types::{
     game_controller_state::GameControllerState, line_data::LineData, primary_state::PrimaryState,
 };
 
+#[derive(Deserialize, Serialize)]
 pub struct LocalizationRecorder {
```
Comment (h3ndrk) [inline crates/control/src/localization_recorder.rs:19]: Remove or track in ideas for next iteration

File: crates/code_generation/src/cyclers.rs (hunk 6 above)
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs:379]: No

File: crates/audio/src/whistle_detection.rs [hunk 11]
```diff
@@ -20,8 +21,11 @@ pub const NUMBER_OF_AUDIO_CHANNELS: usize = 4;
 pub const NUMBER_OF_AUDIO_SAMPLES: usize = 2048;
 const NUMBER_OF_FREQUENCY_SAMPLES: usize = NUMBER_OF_AUDIO_SAMPLES / 2;
 
+#[derive(Deserialize, Serialize)]
 pub struct WhistleDetection {
+    #[serde(skip, default = "deserialize_not_implemented")]
```
Comment (schmidma) [inline crates/audio/src/whistle_detection.rs:26]: what do you think of https://doc.rust-lang.org/std/macro.unimplemented.html instead of `deserialize_not_implemented`?

File: crates/control/src/localization_recorder.rs (hunk 10 above)
Comment (schmidma) [inline crates/control/src/localization_recorder.rs:19]: Can we remove this node now?

File: crates/hardware/src/lib.rs [hunk 12]
```diff
@@ -43,6 +44,11 @@ pub trait PathsInterface {
     fn get_paths(&self) -> Paths;
 }
 
+pub trait RecordingInterface {
+    fn get_recording(&self) -> bool;
+    fn set_recording(&self, enable: bool);
```
Comment (schmidma) [inline crates/hardware/src/lib.rs:49]: As discussed. Please reiterate on the naming here ;)

File: crates/code_generation/src/run.rs [hunk 13]
```diff
@@ -147,6 +199,8 @@ fn generate_cycler_constructors(cyclers: &Cyclers) -> TokenStream {
                 communication_server.get_parameters_reader(),
                 #own_producer_identifier
                 #(#other_cycler_inputs,)*
+                recording_sender.clone(),
+                cycler_instances_to_be_recorded.contains(#cycler_instance_name),
```
Comment (schmidma) [inline crates/code_generation/src/run.rs:203]: Please extract that variable to an earlier position like e.g. `enable_recording`

**Review by schmidma — CHANGES_REQUESTED** [review]
only minor remarks. This looks good to me in general. Let's see a test game before merging this?

File: crates/code_generation/src/cyclers.rs [hunk 14]
```diff
@@ -698,20 +888,32 @@ fn generate_context_initializers(node: &Node, cycler: &Cycler) -> TokenStream {
...
     node.contexts
         .main_outputs
         .iter()
         .filter_map(|field| match field {
             Field::MainOutput { name, .. } => {
+                let error_message = format!("failed to record {name}");
+                let recording_serialization = match recording_generation {
+                    RecordingGeneration::Generate => quote! {
+                        if enable_recording {
+                            bincode::serialize_into(&mut recording_frame, &main_outputs.#name.value).wrap_err(#error_message)?;
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs:910]: I generally agree with optimizing. Do we want to do that in a later iteration? This latency is currently in the range 0.1 milliseconds in the control cycler (the only recorded cycler at the moment). In the future we can further improve the point of recording, e.g. recording only cycler states at the beginning and the rest in the end. But we also need to be careful to finish cycles early but also keep holding locks for the values to record s.t. we record the values of the last cycle...

File: crates/code_generation/src/run.rs [hunk 15]
```diff
@@ -97,6 +113,42 @@ fn generate_future_queues(cyclers: &Cyclers) -> TokenStream {
...
+        let error_message = format!("failed to write into recording file for {instance}");
+        quote! {
+            crate::cyclers::RecordingFrame::#instance_name { data } => #instance_name_snake_case.write_all(data.as_slice()).wrap_err(#error_message)?,
+        }
+    });
+
+    quote! {
+        std::thread::Builder::new()
+            .name("Recording".to_string())
+            .spawn(move || -> color_eyre::Result<()> {
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/run.rs:137]: Yes

```
error[E0282]: type annotations needed
    --> target/incremental/build/behavior_simulator-b25b3ec13813f75b/out/generated_code.rs:1178:17
     |
1178 |                 Ok(())
     |                 ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
     |
help: consider specifying the generic arguments
     |
1178 |                 Ok::<(), E>(())
     |                   +++++++++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `behavior_simulator` (lib test) due to previous error
warning: build failed, waiting for other jobs to finish...
```

File: crates/code_generation/src/run.rs [hunk 16]
```diff
@@ -97,6 +113,42 @@ fn generate_future_queues(cyclers: &Cyclers) -> TokenStream {
...
+                use std::io::Write;
+                let seconds = std::time::SystemTime::now().duration_since(std::time::SystemTime::UNIX_EPOCH).unwrap().as_secs();
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/run.rs:139]: Added to Ideas for Next Iteration

File: crates/audio/src/whistle_detection.rs (hunk 11 above)
Comment (h3ndrk) [reply crates/audio/src/whistle_detection.rs:26]:
```
error[E0423]: expected function, found macro `unimplemented`
  --> crates/audio/src/whistle_detection.rs:26:29
   |
26 |     #[serde(skip, default = "unimplemented")]
   |                             ^^^^^^^^^^^^^^^ not a function
   |
help: use `!` to invoke the macro
   |
26 |     #[serde(skip, default = "unimplemented"!)]
   |                                            +
```

File: crates/control/src/localization_recorder.rs (hunk 10 above)
Comment (h3ndrk) [reply crates/control/src/localization_recorder.rs:19]: Yesterday I added this to Ideas for Next Iteration because the localization recording is a fully working feature (with replay in Twix) but recording alone is not feature-complete since replay is still missing. Once a basic version of replay is working, we can delete it.

File: crates/code_generation/src/cyclers.rs [hunk 17]
```diff
@@ -353,11 +371,13 @@ fn generate_cycle_method(cycler: &Cycler, cyclers: &Cyclers) -> TokenStream {
...
         .setup_nodes
         .iter()
-        .map(|node| generate_node_execution(node, cycler));
+        .map(|node| generate_node_execution(node, cycler, RecordingGeneration::Generate));
     let cycle_node_executions = cycler
         .cycle_nodes
         .iter()
-        .map(|node| generate_node_execution(node, cycler));
+        .map(|node| generate_node_execution(node, cycler, RecordingGeneration::Skip));
+    let required_inputs = get_required_inputs(cycler);
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs:379]: "required inputs" is now "required cross inputs". I hope this is sufficient now, otherwise please make a proposal :wink:

File: crates/code_generation/src/cyclers.rs [hunk 18]
```diff
@@ -427,17 +459,162 @@ fn generate_cycle_method(cycler: &Cycler, cyclers: &Cyclers) -> TokenStream {
...
+                }
             }
             self.own_changed.notify_one();
             Ok(())
         }
     }
 }
 
+fn get_required_inputs(cycler: &Cycler) -> Vec<Field> {
+    let mut required_inputs = Vec::new();
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs:481]: HashSettified now

File: crates/code_generation/src/run.rs [hunk 19]
```diff
@@ -97,6 +113,42 @@ fn generate_future_queues(cyclers: &Cyclers) -> TokenStream {
         .collect()
 }
 
+fn generate_recording_thread(cyclers: &Cyclers) -> TokenStream {
+    let file_creations = cyclers.instances().map(|(_cycler, instance)| {
+        let instance_name_snake_case = format_ident!("{}", instance.to_case(Case::Snake));
+        let instance_name = format!("logs/{instance}.{{seconds}}.bincode");
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/run.rs:119]: It is now "recording file name"

File: crates/code_generation/src/run.rs [hunk 20]
```diff
@@ -97,6 +113,42 @@ fn generate_future_queues(cyclers: &Cyclers) -> TokenStream {
...
+        let error_message = format!("failed to create recording file for {instance}");
+        quote! {
+            let mut #instance_name_snake_case = std::io::BufWriter::new(std::fs::File::create(format!(#instance_name)).wrap_err(#error_message)?);
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/run.rs:122]: Added a TODO comment

File: crates/hardware/src/lib.rs (hunk 12 above)
Comment (h3ndrk) [reply crates/hardware/src/lib.rs:49]: Renamed to

- "should record"
- "set whether to record"

Feel free to propose something :wink:

File: crates/code_generation/src/run.rs (hunk 13 above)
Comment (h3ndrk) [reply crates/code_generation/src/run.rs]: Extracted to variable

File: crates/audio/src/whistle_detection.rs (hunk 11 above)
Comment (h3ndrk) [reply crates/audio/src/whistle_detection.rs:26]: Same with `panic!()`. I thought of something like [`panic_any()`](https://doc.rust-lang.org/std/panic/fn.panic_any.html) function but I need a function `fn() -> T`.

File: crates/code_generation/src/cyclers.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/code_generation/src/cyclers.rs]: I think, the "required" needs context. Required for what?

File: crates/code_generation/src/cyclers.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/code_generation/src/cyclers.rs:918]: Then at least, we should not forget about it? I'm generally concerned on doing that interleaved, but 0.1ms might actually be negligible.

File: crates/audio/src/whistle_detection.rs (hunk 11 above)
Comment (schmidma) [reply crates/audio/src/whistle_detection.rs:26]: Then at least use it in your custom function?

File: crates/code_generation/src/cyclers.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs]: Let's make it longer :joy: How about "cross input required for recording"?

[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs]: Kloppt euch gern :joy:

File: crates/audio/src/whistle_detection.rs (hunk 11 above)
Comment (h3ndrk) [reply crates/audio/src/whistle_detection.rs:26]: Using it now. I thought you wanted to use `unreachable!()` which I would argue against. But `unimplemented!()` is fine for me.

File: crates/code_generation/src/cyclers.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/code_generation/src/cyclers.rs]: It is now `cross_inputs`, I hope this is fine now.

File: crates/hardware/src/lib.rs (hunk 12 above)
Comment (h3ndrk) [reply crates/hardware/src/lib.rs]: To explain my motivation in keeping it for now and trying to shortcut the discussion: I think we should keep in mind the intended maturity level. The long-term plan is (as written in a comment earlier) that the `framework.json`, Communication, and `HardwareInterface` will provide whether they want to record, in which frequency, and/or whether they want to record a burst for a given time duration. This will all be some state that is written once in a cycle or in a lower frequency. The simple boolean that currently exists will be replaced by that. Therefore, I think we should not overfit to it now. In addition, I would vote against making the robotics domain longer due to calling two functions instead of one. The idea of "record this cycle" changes the API and requires additional state and code to reset the state, while we are not yet sure whether this additional code can survive under the planned ideas.

**Review by schmidma — APPROVED** [review]
